crates/typst-kit/src/fonts.rs

```diff
@@ -0,0 +1,138 @@
+use std::path::PathBuf;
+use std::sync::OnceLock;
+use std::{fs, path::Path};
+
+use fontdb::{Database, Source};
+use typst::text::{Font, FontBook, FontInfo};
+use typst_timing::TimingScope;
+
+/// Searches for fonts.
+#[derive(Debug)]
+pub struct FontSearcher {
+    /// Metadata about all discovered fonts.
+    pub book: FontBook,
+    /// Slots that the fonts are loaded into.
+    pub fonts: Vec<FontSlot>,
+}
+
+/// Holds details about the location of a font and lazily the font itself.
+#[derive(Debug)]
+pub struct FontSlot {
+    /// The path at which the font can be found on the system.
+    path: PathBuf,
+    /// The index of the font in its collection. Zero if the path does not point
+    /// to a collection.
+    index: u32,
+    /// The lazily loaded font.
+    font: OnceLock<Option<Font>>,
+}
+
+impl FontSlot {
+    /// Returns the path at which the font can be found on the system.
+    pub fn path(&self) -> &Path {
+        &self.path
```

Comment:
For embedded fonts, the path is empty. So we should maybe switch it to `Option` or at least note it.